transient/config.py

```diff
@@ -0,0 +1,261 @@
+
+    return final_dict
+
+
+def parse_config_file(config_file_path: str) -> MutableMapping[str, Any]:
+    """Parses the given config file and returns the contents as a dictionary
+    """
+    with open(config_file_path) as file:
+        config_file = file.read()
+
+    try:
+        parsed_config_file = toml.loads(config_file)
+    except RuntimeError as error:
+        print(error, file=sys.stderr)
+        sys.exit(1)
+
+    return parsed_config_file
+
+
+def get_line_number_of_string_in_config_file(string: str, config_file_path: str) -> int:
+    """Returns the line number where the config file contains the given string
+    """
+    with open(config_file_path) as config_file:
+        for line_number, line in enumerate(config_file, start=1):
+            if string in line:
+                return line_number
+
+    print("Line not found", file=sys.stderr)
+
+    return -1
+
+
+def print_error_message_for_bad_config_file(
+    error: ValidationError, config_file_path: str
+) -> None:
+    """Prints an error message regarding a bad configuration in the config file
+       and its associated line number
+    """
+    for key in error.messages:
+
+        # Revert the key to its preformatted state
+        key = key.replace("_", "-")
+
+        line_number = get_line_number_of_string_in_config_file(key, config_file_path)
+        print(f"Invalid configuration file, line {line_number}: {error}", file=sys.stderr)
+
+
+def load_config_file(config_file_path: str) -> Config:
+    """Returns the validated schema from the config file
+    """
+    parsed_config_file = parse_config_file(config_file_path)
+
+    reformatted_args = replace_hyphens_with_underscores_in_dict_keys(
+        parsed_config_file["transient"]
+    )
+
+    reformatted_args["qemu_args"] = parsed_config_file["qemu"]["qemu-args"]
+
+    transient_args_schema = TransientConfigSchema()
+
+    try:
+        config: Config = transient_args_schema.load(reformatted_args)
+    except ValidationError as error:
+        print_error_message_for_bad_config_file(error, config_file_path)
+        sys.exit(1)
+
+    return config
+
+
+def create_transient_run_config_from_cli_args(cli_args: Dict[Any, Any]) -> Config:
+    """Creates and validates an Arg dictionary to be used by Transient-run
+       given the CLI arguments and, if specified, a config file
+
+       Note that the CLI arguments take precedence over the config file
+    """
+    try:
+        transient_args_schema = TransientConfigSchema()
+    except ValidationError as error:
+        print("Invalid command line arguments given: {error}", file=sys.stderr)
+        sys.exit(1)
+
+    config: Config = transient_args_schema.load(cli_args)
+
+    if config.config:
+        config_file = load_config_file(config.config)
+
+        # Combine the arguments from the CLI and configuration file
+        # CLI arguments take precedence
+        for key in config_file:
+            if not config[key] and config_file[key]:
```

Comment:
I don't fully follow this. Will it not get confused if `config[key]` exists but is falsey (e.g. if it's 0)

Reply:
I struggled to make this code fragment more readable.

The intended behavior is to assign `config[key]` if it's falsey (None, ( ), [ ], ' ', 0, { }) and `config_file[key]` is NOT falsey.

The purpose of checking if `config[key]` is falsey is to prevent being overwritten by the config file.
The purpose of checking if `config_file[key]` is falsey is to prevent unnecessary falsey assignment.

I was thinking of adding a helper function such as `config_option_is_empty()` and `config_option_is_not_empty()` so:

```python
if config_option_is_empty(config[key]) and config_option_is_not_empty(config_file[key]):
```

Any thoughts? I want to make this as easily comprehensible as possible.

Reply:
I still don't follow. Suppose I set `-shutdown-timeout 0` on the commandline, will this not override that value with something from the config file? I don't understand what the _value_ of the configuration matters, don't we only care about whether an option was specified on the commandline at all? If so, we should always use it, regardless of whether the value the user specified happens to be falsey or not.

Reply:
I'm not sure. We can always just set the 'default' value to a python object we define like NotPassed or something. Then we can compare against that, as a user could not pass that value from the command line.

Reply:
I've tried to prototype the idea of passing 'default' values with a `NotGiven` type. I found out that `Click` type inferences the expected type via the default argument, so I have to use `StringNotGiven`, `IntegerNotGiven`, and `ListNotGiven`.

But there is still a problem. Flags are boolean values and default to `False` if not set, _always_  overwriting the config file value. I think the most intuitive case would be to ignore `False` boolean flags, but this is problematic because `Click` automatically provides both enabling and disabling flags, for example, `-prepare-ssh` and `-no-prepare-ssh`. This would ignore the disable flag.

I'm still thinking through this. Let me know if you have any thoughts.

Reply:
I resolved this by implementing default values that indicate the option was not set on the CLI.

If a flag is not passed on the CLI, the default `False` value does **not** override the value found in the config file.
If a flag is passed on the CLI, it is `True` and will override the value found in the config file.